File: mlp_xor.m
function mlp_xor

    % XOR input and output
    X = [0 0; 0 1; 1 0; 1 1];
    Y = [0; 1; 1; 0];

    % Network configuration
    inputSize = 2; % number of features
    hiddenLayerSizes = [2, 2]; % number of nodes in hidden layers
    outputSize = 1; % number of output nodes
    learnRate = 0.1; % learning rate
    maxEpochs = 20000; % max number of epochs
    epsilon = 1e-3; % error threshold

    % Weights initialization
    W1 = rand(hiddenLayerSizes(1), inputSize) - 0.5;
    b1 = rand(hiddenLayerSizes(1), 1) - 0.5;
    W2 = [];
    b2 = [];
    if length(hiddenLayerSizes) > 1
        W2 = rand(hiddenLayerSizes(2), hiddenLayerSizes(1)) - 0.5;
        b2 = rand(hiddenLayerSizes(2), 1) - 0.5;
    end
    W3 = rand(outputSize, hiddenLayerSizes(end)) - 0.5;
    b3 = rand(outputSize, 1) - 0.5;

    % Training loop
    for epoch = 1:maxEpochs
        for i = 1:size(X, 1)
            % Forward pass
            a1 = sigmoid(W1 * X(i, :)' + b1);
            a2 = [];
            if ~isempty(W2)
                a2 = sigmoid(W2 * a1 + b2);
            end
            if isempty(a2)
                output = sigmoid(W3 * a1 + b3);
            else
                output = sigmoid(W3 * a2 + b3);
            end

            % Error calculation
            error = Y(i) - output;

            % Backpropagation
            delta3 = error .* sigmoid_prime(output);
            delta2 = [];
            if ~isempty(W2)
                delta2 = (W3' * delta3) .* sigmoid_prime(a2);
            end
            delta1 = [];
            if isempty(delta2)
                delta1 = (W3' * delta3) .* sigmoid_prime(a1);
            else
                delta1 = (W2' * delta2) .* sigmoid_prime(a1);
            end

            % Weights update
            if ~isempty(delta2)
                W3 = W3 + learnRate * delta3 * a2';
                b3 = b3 + learnRate * delta3;
                W2 = W2 + learnRate * delta2 * a1';
                b2 = b2 + learnRate * delta2;
            else
                W3 = W3 + learnRate * delta3 * a1';
                b3 = b3 + learnRate * delta3;
            end
            W1 = W1 + learnRate * delta1 * X(i, :);
            b1 = b1 + learnRate * delta1;

        end

        % Display training progress
        if mod(epoch, 1000) == 0
            disp(['Epoch: ' num2str(epoch) ' Error: ' num2str(sum(abs(error)))]);
        end

        % Check convergence
        if sum(abs(error)) < epsilon
            disp(['Converged at epoch: ' num2str(epoch)]);
            break;
        end
    end

    % Testing
    for i = 1:size(X, 1)
        a1 = sigmoid(W1 * X(i, :)' + b1);
        a2 = [];
        if ~isempty(W2)
            a2 = sigmoid(W2 * a1 + b2);
        end
        if isempty(a2)
            output = sigmoid(W3 * a1 + b3);
        else
            output = sigmoid(W3 * a2 + b3);
        end
        disp(['Input: ' num2str(X(i, :)) ' Expected: ' num2str(Y(i)) ' Output: ' num2str(round(output))]);
    end
end

% Sigmoid activation function
function s = sigmoid(z)
    s = 1 ./ (1 + exp(-z));
end

% Derivative of the sigmoid function
function s_prime = sigmoid_prime(z)
    s_prime = z .* (1 - z);
end

File: MLP_XOR.py
#import required libaries
import numpy as np
from matplotlib import pyplot as plt

class MLP():
    
    """
    This is the MLP class used to feedforward and backpropagate the network across a defined number 
    of iterations and produce predictions. After iteration the predictions are assessed using 
    Binary Cross Entropy Cost function.  
    """
    
    print('Running...')
    
    def __init__(self, design_matrix, Y, iterations=100000, lr=1e-1, input_layer = 2, hidden_layer = 3,output_layer =1):
        self.design_matrix = design_matrix #design matrix attibute
        self.iterations = iterations #iterations attibute
        self.lr = lr #learning rate attibute
        self.input_layer = input_layer #input layer attibute 
        self.hidden_layer = hidden_layer #hidden layer attibute
        self.output_layer = output_layer #output layer attibute
        self.weight_matrix_1 = np.random.randn(self.input_layer, self.hidden_layer) #weight attribute connecting to the hidden layer
        self.weight_matrix_2 = np.random.randn(self.hidden_layer, self.output_layer)#weight attribute connecting to the output layer
        self.cost = [] #cost list attribute 
        self.p_hats = [] #predictions list attribute

    def sigmoid(self, x): # sigmoid function used at the hidden layer and output layer
        return 1 / (1 + np.exp(-x))

    def sigmoid_derivative(self, x): # sigmoid derivative used for backpropgation 
        return self.sigmoid(x) * (1 - self.sigmoid(x))

    def forward_propagation(self):#define function to feedforward the network 
        z = np.dot(self.design_matrix, self.weight_matrix_1) #linear transformation to the hidden layer
        activation_func = self.sigmoid(z)#hidden layer activation function
        zh = np.dot(activation_func, self.weight_matrix_2)#linear transformation to the output layer
        p_hat = self.sigmoid(zh)#output layer prediction
        return z, activation_func, zh, p_hat

    def BCECost(self, y, p_hat): # binary cross entropy cost function
        bce_cost = -(np.sum(y * np.log(p_hat) + (1 - y) * np.log(1 - p_hat))) / len(y)
        return bce_cost

    def backword_prop(self, z_1, activation_func, z_2, p_hat): #backpropagation
        del_2_1 = p_hat - Y
        partial_deriv_2 = np.dot(activation_func.T, del_2_1) #∂loss/∂p *∂p/∂zh * ∂zh/∂wh
        del_1_1 = del_2_1 
        del_1_2 = np.multiply(del_1_1, self.weight_matrix_2.T) 
        del_1_3 = np.multiply(del_1_2, self.sigmoid_derivative(z_1))
        partial_deriv_1 = np.dot(self.design_matrix.T, del_1_3) #∂loss/∂p * ∂p/∂zh * ∂zh/∂h * ∂h/∂z * ∂z/∂w
        return partial_deriv_2, partial_deriv_1

    def train(self):#train the network
        for i in range(self.iterations): #loop based on number of iterations
            z_1, activation_func, z_2, p_hat = self.forward_propagation()# feedforward
            partial_deriv_2, partial_deriv_1 = self.backword_prop(z_1, activation_func, z_2, p_hat)#backpropgate
            self.weight_matrix_1 = self.weight_matrix_1 - self.lr * partial_deriv_1#update weights connecting to the hidden layer (gradient descent)
            self.weight_matrix_2 = self.weight_matrix_2 - self.lr * partial_deriv_2#update weights connecting to the output layer (gradient descent )
            self.cost.append(self.BCECost(Y, p_hat))#store BCE cost in list
            self.p_hats.append(p_hat)#store predictions in list
        
        
        print('Training Complete')
        print('----------------------------------------------------------------------------')

# Prepare the XOR Logic Gate data: create an array for each training example x feature, and an array for each corrosponding y label.
X = np.array([[1, 0], [0, 1], [0, 0], [1, 1]]) #input features (4 x 2 design matrix)
Y = np.array([[1], [1], [0], [0]])#ground truth y labels (4x1)

mlp = MLP(X,Y)#Pass data to the model (design matrix and y label)
mlp.train() #Train the model

#plot the cost function
plt.grid()
plt.plot(range(mlp.iterations),mlp.cost)
plt.xlabel('Iterations')
plt.ylabel('Cost')
plt.title('BCE Cost Function')

#Print predictions, number of iterations and the ground truth labels.
print(f'\n The MLP predictions for each training example, based on {mlp.iterations} iterations are:\n\n{np.round(mlp.p_hats[-1],2)}')
print('\n----------------------------------------------------------------------------')
print(f'\n The ground truth Y labels are are:\n\n{Y}')